src/service.rs

```diff
@@ -305,37 +310,148 @@ where
     for<'a> <TYPES::SignatureKey as TryFrom<&'a TaggedBase64>>::Error: Display,
 {
     /// TODO: fetch actual blocks
-    async fn bundle(&self, _view_number: u64) -> Result<Bundle<TYPES>, BuildError> {
-        const FEE_AMOUNT: u64 = 1;
+    async fn bundle(&self, view_number: u64) -> Result<Bundle<TYPES>, BuildError> {
+        let start = Instant::now();
+        let requested_view = TYPES::Time::new(view_number);
+        // TODO: We should either change the API to match leader's intent or track the builder states
+        // differently, as this won't work for a view immediately after a view failure.
+        let parent_view = TYPES::Time::new(view_number.saturating_sub(1));
+
+        loop {
+            // Couldn't serve a bundle in time
+            if start.elapsed() > self.api_timeout {
+                tracing::warn!(?requested_view, "Timeout while trying to serve a bundle");
+                return Err(BuildError::NotFound);
+            };
+
+            // Look up a builder for this view.
+            //
+            // TODO: We currently choose the first one matching the view we found,
+            // we should introduce parent commitment to the API to match the leader's intent
+            let Some((state_id, sender)) = self
+                .global_state
+                .read_arc()
+                .await
+                .spawned_builder_states
+                .iter()
+                .find(|(id, _)| id.view == parent_view)
+                .map(|(id, sender)| (id.clone(), sender.clone()))
+            else {
+                let global_state = self.global_state.read_arc().await;
+                if requested_view <= global_state.last_garbage_collected_view_num
+                    && global_state.highest_view_num_builder_id.view
+                        != global_state.last_garbage_collected_view_num
```

Comment:
Why do we need this comparison in addition to the first one?

Reply:
Added some comments explaining this